File: Lab/SEM4/MPP/lab4/moto_c/repository/RacerRepositoryImpl.cs
using moto_c.entity;
using MySqlConnector;
using NLog;

namespace moto_c.repository;

public class RacerRepositoryImpl:RacerRepository
{
    private readonly MySqlConnection _connection = DatabaseConnection.getInstance().GetConnection();
    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
    private readonly RaceEventRepository _raceEventRepository = RaceEventRepositoryImpl.getInstance();
    private static RacerRepositoryImpl? instance;

    private RacerRepositoryImpl() {}

    public static RacerRepositoryImpl getInstance()
    {
        if (instance == null)
            instance = new RacerRepositoryImpl();
        return instance;
    }
    
    public Racer? find(long id)
    {
        logger.Trace("Entry Find id={}", id);
        string sql = "SELECT * FROM racers WHERE id = ? LIMIT 1";

        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@id", id);

                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    if (rs.Read())
                    {
                        string name = rs.GetString("name");
                        string cnp = rs.GetString("cnp");
                        RaceEvent? engine = _raceEventRepository.find(rs.GetInt32("engine"));
                        Team team = Enum.Parse<Team>(rs.GetString("team"));

                        Racer racer = new Racer(name, cnp, team, engine!);
                        racer.id = id;

                        logger.Trace("Exit Find id={}", id);
                        return racer;
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }

        logger.Trace("Exit Find id={}", id);
        return null;
    }

    public List<Racer> getAll()
    {
        logger.Trace("Entry GetAll");
        string sql = "SELECT * FROM racers";
        List<Racer> racers = new List<Racer>();

        try
        {
            using (MySqlCommand command = new MySqlCommand(sql, _connection))
            {
                using (MySqlDataReader rs = command.ExecuteReader())
                {
                    while (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        string name = rs.GetString("name");
                        string cnp = rs.GetString("cnp");
                        RaceEvent? engine = _raceEventRepository.find(rs.GetInt32("engine"));
                        Team team = Enum.Parse<Team>(rs.GetString("team"));

                        Racer racer = new Racer(name, cnp, team, engine!);
                        racer.id = id;
                        
                        racers.Add(racer);
                    }
                }
            }
        }
        catch (MySqlException ex)
        {
            logger.Error(ex.Message);
            Console.Error.WriteLine(ex.Message);
        }

        logger.Trace("Exit GetAll");
        return racers;
    }

    public List<Racer> findBy(Team team)
    {
        logger.Trace("Entry FindBy team={}", team);
        string sql = "SELECT * FROM racers WHERE team = (@team)";
        List<Racer> racers = new List<Racer>();

        try
        {
            using (MySqlCommand command = new MySqlCommand(sql, _connection))
            {
                command.Parameters.AddWithValue("@team", team.ToString());
                using (MySqlDataReader rs = command.ExecuteReader())
                {
                    while (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        string name = rs.GetString("name");
                        string cnp = rs.GetString("cnp");
                        RaceEvent? engine = _raceEventRepository.find(rs.GetInt32("engine"));

                        Racer racer = new Racer(name, cnp, team, engine!);
                        racer.id = id;
                        
                        racers.Add(racer);
                    }
                }
            }
        }
        catch (MySqlException ex)
        {
            logger.Error(ex.Message);
            Console.Error.WriteLine(ex.Message);
        }

        logger.Trace("Exit FindBy team={}", team);
        return racers;
    }

    public List<Racer> findBy(RaceEvent engine)
    {
        logger.Trace("Entry FindBy engine={}", engine.engine);
        string sql = "SELECT * FROM racers WHERE engine=(@engine)";
        List<Racer> racers = new List<Racer>();

        try
        {
            using (MySqlCommand command = new MySqlCommand(sql, _connection))
            {
                command.Parameters.AddWithValue("@engine", engine.engine);
                using (MySqlDataReader rs = command.ExecuteReader())
                {
                    while (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        string name = rs.GetString("name");
                        string cnp = rs.GetString("cnp");
                        Team team = Enum.Parse<Team>(rs.GetString("team"));

                        Racer racer = new Racer(name, cnp, team, engine);
                        racer.id = id;
                        
                        racers.Add(racer);
                    }
                }
            }
        }
        catch (MySqlException ex)
        {
            logger.Error(ex.Message);
            Console.Error.WriteLine(ex.Message);
        }

        logger.Trace("Exit FindBy engine={}", engine.engine);
        return racers;
    }

    public void add(Racer racer)
    {
        logger.Trace("Entry Add racer={}", racer.name);
        string sql = "INSERT INTO racers (name,cnp,engine, team) VALUES (@name,@cnp,@engine,@team)";

        try
        {
            using (MySqlCommand command = new MySqlCommand(sql, _connection))
            {
                command.Parameters.AddWithValue("@name", racer.name);
                command.Parameters.AddWithValue("@cnp", racer.cnp);
                command.Parameters.AddWithValue("@engine", racer.engine.engine);
                command.Parameters.AddWithValue("@team", racer.team.ToString());
                
                int result = command.ExecuteNonQuery();
                logger.Trace("Saved {} instance", result);
            }
        }
        catch (MySqlException ex)
        {
            logger.Error(ex.Message);
            Console.Error.WriteLine(ex.Message);
        }
        
        logger.Trace("Exit Add");
    }

    public void modify(Racer racer)
    {
        logger.Trace("Entry Modify racer id={}", racer.id);
        string sql = "UPDATE racers SET name=@name, cnp=@cnp, engine=@engine, team=@team WHERE id=@id";

        try
        {
            using (MySqlCommand command = new MySqlCommand(sql, _connection))
            {
                command.Parameters.AddWithValue("@name", racer.name);
                command.Parameters.AddWithValue("@cnp", racer.cnp);
                command.Parameters.AddWithValue("@engine", racer.engine.engine);
                command.Parameters.AddWithValue("@team", racer.team.ToString());
                command.Parameters.AddWithValue("@id", racer.id);
                
                command.ExecuteNonQuery();
            }
        }
        catch (MySqlException ex)
        {
            logger.Error(ex.Message);
            Console.Error.WriteLine(ex.Message);
        }
        
        logger.Trace("Exit Modify");
    }
}

File: Lab/SEM4/MPP/lab4/moto_java/src/main/java/org/zdroba/repository/RacerRepositoryImpl.java
package org.zdroba.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.zdroba.entity.RaceEvent;
import org.zdroba.entity.Racer;
import org.zdroba.entity.Team;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class RacerRepositoryImpl implements RacerRepository{

    private static RacerRepositoryImpl instance;

    private RacerRepositoryImpl() {}

    public static RacerRepositoryImpl getInstance() {
        if (instance == null)
            instance = new RacerRepositoryImpl();
        return instance;
    }

    private final RaceEventRepository eventRepository = RaceEventRepositoryImpl.getInstance();
    private final Connection connection = DataBaseConnection.getInstance().getConnection();
    private static final Logger logger = LogManager.getLogger();

    @Override
    public Racer find(Long id) {
        logger.traceEntry();
        String sql = "SELECT * FROM racers WHERE id = ?";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setLong(1,id);

            try(ResultSet rs = ps.executeQuery()){

                if(rs.next()){
                    String name = rs.getString("name");
                    String cnp = rs.getString("cnp");
                    RaceEvent engine = eventRepository.find(rs.getInt("engine"));
                    Team team = Team.valueOf(rs.getString("team"));

                    Racer racer = new Racer(name,cnp,team,engine);
                    racer.setId(id);

                    logger.traceExit();
                    return racer;
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return null;
    }

    @Override
    public List<Racer> getAll() {
        logger.traceEntry();
        String sql = "SELECT * FROM racers";

        List<Racer> racers = new ArrayList<>();

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            try(ResultSet rs = ps.executeQuery()){

                while(rs.next()){
                    Long id = rs.getLong("id");
                    String name = rs.getString("name");
                    String cnp = rs.getString("cnp");
                    RaceEvent engine = eventRepository.find(rs.getInt("engine"));
                    Team team = Team.valueOf(rs.getString("team"));

                    Racer racer = new Racer(name,cnp,team,engine);
                    racer.setId(id);

                    racers.add(racer);
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return racers;
    }

    @Override
    public List<Racer> getBy(Team team) {
        logger.traceEntry();
        String sql = "SELECT * FROM racers WHERE team = ?";

        List<Racer> racers = new ArrayList<>();

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setString(1, team.name());

            try(ResultSet rs = ps.executeQuery()){

                while(rs.next()){
                    Long id = rs.getLong("id");
                    String name = rs.getString("name");
                    String cnp = rs.getString("cnp");
                    RaceEvent engine = eventRepository.find(rs.getInt("engine"));

                    Racer racer = new Racer(name,cnp,team,engine);
                    racer.setId(id);

                    racers.add(racer);
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return racers;
    }

    @Override
    public List<Racer> getBy(RaceEvent engine) {
        logger.traceEntry();
        String sql = "SELECT * FROM racers WHERE engine = ?";

        List<Racer> racers = new ArrayList<>();

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setInt(1, engine.getEngine());

            try(ResultSet rs = ps.executeQuery()){

                while(rs.next()){
                    Long id = rs.getLong("id");
                    String name = rs.getString("name");
                    String cnp = rs.getString("cnp");
                    Team team = Team.valueOf(rs.getString("team"));

                    Racer racer = new Racer(name,cnp,team,engine);
                    racer.setId(id);

                    racers.add(racer);
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return racers;
    }

    @Override
    public void add(Racer racer) {
        logger.traceEntry("saving Racer{}", racer);
        String sql = "INSERT INTO racers (name,cnp,engine,team) VALUES (?,?,?,?)";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setString(1, racer.getName());
            ps.setString(2, racer.getCnp());
            ps.setInt(3, racer.getEngine().getEngine());
            ps.setString(4, racer.getTeam().name());

            int result = ps.executeUpdate();
            logger.trace("Saved {} instance", result);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
    }

    @Override
    public void modify(Racer racer) {
        logger.traceEntry("Modifying Racer{}", racer.getId());
        String sql = "UPDATE racers SET name = ?, cnp = ?, team = ?, engine = ? WHERE id = ?";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setString(1, racer.getName());
            ps.setString(2, racer.getCnp());
            ps.setString(3, racer.getTeam().name());
            ps.setLong(4, racer.getEngine().getId());
            ps.setLong(5, racer.getId());

            int result = ps.executeUpdate();
            logger.trace("Updated {} instance", result);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
    }
}
